src/group/managed_group/mod.rs

```diff
@@ -0,0 +1,276 @@
+///
+/// The API is modelled such that it can serve as a direct interface to the Delivery Serive. Functions that modify the public state
+/// of the group will return a `Vec<MLSMessage>` that can be sent to the Delivery Service directly. Conversely, incoming messages
+/// from the Delivery Service can be fed into `process_nessage()`.
+///
+/// A `ManagedGroup` has an internal queue of pending proposals that builds up as new messages are processed. When creating proposals,
+/// those messages are not automatically appended to this queue, instead they have be processed again through `process_message()`.
+/// This allows the Delivery Service to reject them (e.g. if they reference the wrong epoch).
+///
+/// If incoming messages or applied operations are semantically or syntactily incorrect, the function will return a corresponding
+/// error message and the state of the group will remain unchanged.
+///
+/// The application policy for the group can be enforced by implementing the validator callback functions and selectively allowing/
+/// disallowing each operation (see `ManagedGroupCallbacks`)
+///
+/// Changes to the group state are dispatched as events through callback functions (see ManagedGroupCallbacks).
+pub struct ManagedGroup {
+    managed_group_config: ManagedGroupConfig,
+    group: MlsGroup,
+    proposal_queue: ProposalQueue,
+    own_kpbs: Vec<KeyPackageBundle>,
+}
+
+impl ManagedGroup {
+    // === Group creation ===
+
+    /// Creates a new group from scratch with only the creator as a member.
+    pub fn new(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
+        ciphersuite_name: CiphersuiteName,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, GroupError> {
+        let group = MlsGroup::new(
+            &group_id.as_slice(),
+            ciphersuite_name,
+            key_package_bundle,
+            GroupConfig::default(),
+        )?;
+
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group from a `Welcome` message
+    pub fn new_from_welcome(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        welcome: Welcome,
+        ratchet_tree: Option<Vec<Option<Node>>>,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, WelcomeError> {
+        let group = MlsGroup::new_from_welcome(welcome, ratchet_tree, key_package_bundle)?;
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group with the creator and additional members
+    pub fn new_with_members(
```

Comment:
We could probably skip this in a first version, or just make it call `new` and `add_members` (so the API should be the same as `new` and `add_members` combined).